File: lecture_notes/10 refactoring/7-clear.js
'use strict';

const shift = (offset) => (dot) => {
  const moved = {
    x: dot.x += offset.dx,
    y: dot.y += offset.dy,
  };
return moved;
};

const parsers = { string: JSON.parse, object: (obj) => obj };

const conditionalParse = (item) => {
  const parser = parsers[typeof item];
  return parser(item);
};

const polyline = [
  { x: 0, y: 0 },
  { x: 10, y: 10 },
  '{ "x": 20, "y": 20 }',
  { x: 30, y: 30 },
];

const doTask = () => {
  const move = shift({ dx: 10, dy: -5 });
  const parsed = polyline.map(conditionalParse);
  const path = parsed.map(move);
  console.log({ path });
};

return doTask.call();

File: lecture_notes/10 refactoring/7-clear.py
import json

def move(offset):
    return lambda point: {"x": point["x"] + offset["x"], "y": point["y"] + offset["y"]}


parsers = {
    str: json.loads,
    dict: lambda obj: obj}


def conditionalParse(item):
    parser = parsers[type(item)]
    return parser(item)


polyline = [
  { "x": 0, "y": 0 },
  { "x": 10, "y": 10 },
  '{ "x": 20, "y": 20 }',
  { "x": 30, "y": 30 }
]

parsed = list(map(conditionalParse, polyline))
path = list(map(move({ 'x': 10, 'y': -5 }), parsed))
print(path)
